File: core/period/src/main/kotlin/tech/mmarca/openvitals/core/period/PeriodLoadQuery.kt
package tech.mmarca.openvitals.core.period

import java.time.LocalDate

data class PeriodWindows(
    val current: DatePeriod,
    val previous: DatePeriod,
    val baseline: DatePeriod,
)

data class PeriodLoadQuery(
    val range: TimeRange,
    val anchorDate: LocalDate,
    val today: LocalDate = LocalDate.now(),
    val baselineDays: Long = DefaultBaselineDays,
    val weekPeriodMode: WeekPeriodMode = WeekPeriodMode.MONDAY_TO_SUNDAY,
) {
    val selectedDate: LocalDate = anchorDate.coerceAtMost(today)
    val windows: PeriodWindows = periodWindowsFor(
        range = range,
        anchorDate = selectedDate,
        today = today,
        baselineDays = baselineDays,
        weekPeriodMode = weekPeriodMode,
    )
}

class PeriodSelectionDriver(
    initialRange: TimeRange,
    initialDate: LocalDate = LocalDate.now(),
    initialWeekPeriodMode: WeekPeriodMode = WeekPeriodMode.MONDAY_TO_SUNDAY,
    private val onRangeSelected: (TimeRange) -> Unit = {},
) {
    var weekPeriodMode: WeekPeriodMode = initialWeekPeriodMode
    var selection: PeriodSelection = PeriodSelection(
        selectedRange = initialRange,
        selectedDate = initialDate.coerceAtMost(LocalDate.now()),
    )
        private set

    fun selectRange(range: TimeRange): PeriodSelection {
        onRangeSelected(range)
        return update(selection.selectRange(range))
    }

    fun previousPeriod(): PeriodSelection =
        update(selection.previousPeriod())

    fun nextPeriod(): PeriodSelection? {
        val next = selection.nextPeriod(weekPeriodMode = weekPeriodMode)
        return if (next == selection) null else update(next)
    }

    fun selectDate(date: LocalDate): PeriodSelection =
        update(selection.selectDate(date))

    private fun update(next: PeriodSelection): PeriodSelection {
        selection = next
        return next
    }
}

fun periodWindowsFor(
    range: TimeRange,
    anchorDate: LocalDate,
    today: LocalDate = LocalDate.now(),
    baselineDays: Long = DefaultBaselineDays,
    weekPeriodMode: WeekPeriodMode = WeekPeriodMode.MONDAY_TO_SUNDAY,
): PeriodWindows {
    val current = periodFor(range, anchorDate.coerceAtMost(today), today, weekPeriodMode)
    val previous = previousPeriodFor(range, anchorDate, today, weekPeriodMode)
    val baseline = baselinePeriodBefore(current, baselineDays)
    return PeriodWindows(
        current = current,
        previous = previous,
        baseline = baseline,
    )
}

private const val DefaultBaselineDays = 90L

File: lib/core/period/period_load_query.dart
import '../time/local_date.dart';
import 'period_calculations.dart';
import 'time_range.dart';

class PeriodWindows {
  const PeriodWindows({
    required this.current,
    required this.previous,
    required this.baseline,
  });

  final DatePeriod current;
  final DatePeriod previous;
  final DatePeriod baseline;

  @override
  bool operator ==(Object other) =>
      other is PeriodWindows &&
      other.current == current &&
      other.previous == previous &&
      other.baseline == baseline;

  @override
  int get hashCode => Object.hash(current, previous, baseline);
}

class PeriodLoadQuery {
  PeriodLoadQuery({
    required this.range,
    required this.anchorDate,
    LocalDate? today,
    this.baselineDays = defaultBaselineDays,
    this.weekPeriodMode = WeekPeriodMode.mondayToSunday,
  }) : today = today ?? LocalDate.now();

  final TimeRange range;
  final LocalDate anchorDate;
  final LocalDate today;
  final int baselineDays;
  final WeekPeriodMode weekPeriodMode;

  LocalDate get selectedDate => anchorDate.coerceAtMost(today);

  PeriodWindows get windows => periodWindowsFor(
        range,
        selectedDate,
        today: today,
        baselineDays: baselineDays,
        weekPeriodMode: weekPeriodMode,
      );

  @override
  bool operator ==(Object other) =>
      other is PeriodLoadQuery &&
      other.range == range &&
      other.anchorDate == anchorDate &&
      other.today == today &&
      other.baselineDays == baselineDays &&
      other.weekPeriodMode == weekPeriodMode;

  @override
  int get hashCode =>
      Object.hash(range, anchorDate, today, baselineDays, weekPeriodMode);
}
